docs/_docs/20-fasterkv-basics.md

```diff
@@ -109,6 +109,74 @@ Apart from Key and Value, the IFunctions interface is defined on three additiona
+
+**RMW** (`RMWInfo`)
+  - `InPlaceUpdater`
+    - ***Expiration***: The app sets rmwInfo.DeleteRecord true and returns false. FASTER will handle this as if the record was not found, and will enter the `NeedInitialUpdater` path, unless ***Cancellation*** was also done; in that case, FASTER will set recordInfo.Tombstone to true and will return (status.NotFound | status.InPlaceUpdated).
+      - The app may do its own internal "expire and reset" logic within the `InPlaceUpdater` call and return true, which to FASTER is indistinguishable from any other successful `InPlaceUpdater` call.
+    - ***Cancellation***: The app sets rmwInfo.CancelOperation True and returns false. If Expiration was also done, RMW() returns (status.NotFound | status.InPlaceUpdated); otherwise, RMW() returns (status.Canceled).
+    - ***Other false return***: Goes to `InitialUpdater` path.
+  - *InitialUpdater*
+    - `NeedInitialUpdate`
+      - ***Expiration***: Irrelevant because there is no existing record.
+      - ***Cancellation***: The app sets rmwInfo.CancelOperation True and returns false. RMW() returns (status.Canceled).
+      - ***Other false return***: Returns (status.NotFound).
+    - `InitialUpdater`
+      - ***Expiration***: Irrelevant because there is no existing record.
+      - ***Cancellation***: The app sets rmwInfo.CancelOperation True and returns false.  If Expiration was also done, RMW() returns (status.NotFound | status.CreatedRecord); otherwise, RMW() returns (status.Canceled).
+      - ***Other false return***: Returns (status.NotFound) (but no status.Recored.Created).
+    - `PostInitialUpdater`
+      - This is void because the record has already been CAS'd into the hash table, so it must perform operations that do not fail.
+  - *CopyUpdater*
+    - `NeedCopyUpdate`
+      - ***Expiration***: The app sets rmwInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by setting the record's Tombstone and continuing with the CAS, and RMW will return (status.NotFound && status.Record.CreatedRecord).
+      - ***Cancellation***: The app sets rmwInfo.CancelOperation True and returns false. RMW() returns (status.Canceled).
+      - ***Other false return***: Returns (status.Found).
+    - `CopyUpdater`
+      - ***Expiration***: The app sets rmwInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by setting the record's Tombstone and continuing with the CAS, and RMW will return (status.NotFound && status.Record.CreatedRecord). This may be superseded by ***Cancellation***.
+      - ***Cancellation***: The app sets rmwInfo.CancelOperation True and returns false. RMW() returns (status.Canceled). This supersedes ***Expiration***.
+      - ***Other false return***: Returns (status.Found) (but no status.Recored.Created).
+    - `PostCopyUpdater`
+      - This is void because the record has already been CAS'd into the hash table, so it must perform operations that do not fail.
+
```

Comment:
When an RMW operation encounters expiration (without cancellation), we need to retry_now the operation. While this most likely implies a NIU -> IU codepath, it does not necessarily imply so. If another thread managed to add a new value immediately after Tombstoning it, then our retry_now may end up doing a IPU or even a CU. This comment is really to confirm that this interpretation is correct.